ocrd_tesserocr/common.py

```diff
@@ -5,43 +5,145 @@
+    determine the background from the median color (instead of
+    white).
+    
+    Return a new PIL.Image.
+    """
+    # todo: perhaps we should issue a warning if we encounter this
+    # (It should be invalid in PAGE-XML to extend beyond parents.)
+    if not box:
+        box = (0, 0, image.width, image.height)
+    xywh = xywh_from_bbox(*box)
+    background = ImageStat.Stat(image).median[0]
+    new_image = Image.new(image.mode, (xywh['w'], xywh['h']),
+                          background) # or 'white'
+    new_image.paste(image, (-xywh['x'], -xywh['y']))
+    return new_image
+
+# to be refactored into core (as function in ocrd_utils):
+def rotate_coordinates(polygon, angle, orig=np.array([0, 0])):
+    """Apply a passive rotation transformation to the given coordinates.
+    
+    Given a numpy array `polygon` of points and a rotation `angle`,
+    as well as a numpy array `orig` of the center of rotation,
+    calculate the coordinate transform corresponding to the rotation
+    of the underlying image by `angle` degrees at `center` by
+    applying translation to the center, inverse rotation,
+    and translation from the center.
+
+    Return a numpy array of the resulting polygon.
+    """
     angle = np.deg2rad(angle)
+    cos = np.cos(angle)
+    sin = np.sin(angle)
     # active rotation:  [[cos, -sin], [sin, cos]]
     # passive rotation: [[cos, sin], [-sin, cos]] (inverse)
-    return [(orig['x']
-             + (x - orig['x'])*np.cos(angle)
-             + (y - orig['y'])*np.sin(angle),
-             orig['y']
-             - (x - orig['x'])*np.sin(angle)
-             + (y - orig['y'])*np.cos(angle))
-            for x, y in coordinates]
+    return orig + np.dot(polygon - orig, np.array([[cos, sin], [-sin, cos]]).transpose())
+
+# to be refactored into core (as method of ocrd.workspace.Workspace):
+def coordinates_of_segment(segment, parent_image, parent_xywh):
+    """Extract the relative coordinates polygon of a PAGE segment element.
+    
+    Given a Region / TextLine / Word / Glyph `segment` and
+    the PIL.Image of its parent Page / Region / TextLine / Word
+    along with its bounding box, calculate the relative coordinates
+    of the segment within the image. That is, shift all points from
+    the offset of the parent, and (in case the parent was rotated,)
+    rotate all points with the center of the image as origin.
+    
+    Return the rounded numpy array of the resulting polygon.
+    """
+    # get polygon:
+    polygon = np.array(polygon_from_points(segment.get_Coords().points))
+    # offset correction (shift coordinates to base of segment):
+    polygon -= np.array([parent_xywh['x'], parent_xywh['y']])
+    # angle correction (rotate coordinates if image has been rotated):
+    if 'angle' in parent_xywh:
+        polygon = rotate_coordinates(
+            polygon, parent_xywh['angle'],
+            orig=np.array([0.5 * parent_image.width,
```

Comment:
Pls. apologize: I do not understand the constant here...

Reply:
It is rotated around the middle point, hence the `0.5 * width/height`, right?

Reply:
Exactly. The center of rotation for the image (which we have to compensate for when rotating the relative coordinates as well) was half-way into the image.